prow/plugins/override/override.go

```diff
@@ -352,6 +386,13 @@ func handle(oc overrideClient, log *logrus.Entry, e *github.GenericCommentEvent,
 
 		contexts.Insert(status.Context)
 
+		for _, cr := range checkrunContexts {
```

Comment:
Wait, so we are creating a status context with the same name as the checkrun? Are you sure that works?

Reply:
hmm, not sure

I think invoking override should send a request to tide to override the specified context

Reply:
I think I get it now. brb

Reply:
It is not possible.

```
 REDACTED  MCW0CDP3YY  ~  Desktop  Git  $  curl -X PATCH  -H "Authorization:Bearer $TOKEN"   https://api.github.com/repos/borg-land/build-image/check-runs/7660510380   -d '{"conclusion:"neutral"}'
{
  "message": "Invalid app_id `15368` - check run can only be modified by the GitHub App that created it.",
  "documentation_url": "https://docs.github.com/rest/reference/checks#update-a-check-run"
}
```

https://github.com/probot/probot/issues/1065#issuecomment-576444183

Is there another approach?

Reply:
Not sure, it is also possible to write to check runs, but I don't know if GitHub also prevents that if the app wasn't the original author of the check run

Reply:
I have a solution:

https://github.com/borg-land/build-image/pull/1/checks?check_run_id=7669784308

1. Everytime an override is issued, we check if the override is part of the statuses or checkruns. If it a status, then nothing changes
2. If the override is issued for a checkrun, override plugin creates a new checkrun called prow-override and adds the overriden checkruns
```
curl -XPOST -H "Accept: application/vnd.github+json" \ 
  -H "Authorization:Bearer $TOKEN" \
  https://api.github.com/repos/borg-land/build-image/check-runs \
  -d '{"name":"prow-override","head_sha":"4498db721b3262e3b431fc3d6929891b942ab04c","status":"in_progress","output":{"title":"Prow Override","summary":"Prow has received override for the checks below.","text":"{\"override_checkruns\":[\"foo\",\"bar\"]}"}}'

```
3. Tide looks for the prow-override checkrun, reads it and looks for checkruns to ignore when calculating if a PR can be merged

Reply:
hm that works but would add quite a bit of complexity. @chaodaiG and @cjwagner WDYT?

Reply:
One thing that I remember is that it is possible to post multiple results for the same checkrun in which case Tide will select the best: https://github.com/kubernetes/test-infra/blob/84488f7d3087a6b7321448a62e298051b816210e/prow/tide/tide.go#L2307-L2320

Maybe we can use that mechanism? If possible it would be great if we found a solution that doesn't require Tide changes

Reply:
> hm that works but would add quite a bit of complexity. @chaodaiG and @cjwagner WDYT?

Yeah, it also could be confusing from the user's perspective. I'd prefer Alvaro's suggestion if that is viable.

Reply:
Ok

https://api.github.com/repos/borg-land/build-image/commits/4498db721b3262e3b431fc3d6929891b942ab04c/check-runs

Reply:
Were you able to test using the github api if it actually does work?

Reply:
Should be or does? That deduplication logic already exist in Tide so assuming things work that test PR should get merged, can you verify that so we are sure this works end to end?

Reply:
Prow isn't installed on that GitHub Organization. Let me play around with the knative prow GitHub app

Reply:
https://github.com/knative-sandbox/discovery/pull/283

```
# Failing check
 REDACTED  MCW0CDP3YY  ~  $  gh api repos/knative-sandbox/discovery/commits/8db7d8dd315cde9b446aece6175d914b010b841c/check-runs | jq '.check_runs | .[] |select(.name=="analyze / Analyze") '
{
  "id": 7692116478,
  "name": "analyze / Analyze",
  "node_id": "CR_kwDOEGSYGs8AAAABynxh_g",
  "head_sha": "8db7d8dd315cde9b446aece6175d914b010b841c",
  "external_id": "91148dad-8242-58a1-d104-2874ea6ecbb1",
  "url": "https://api.github.com/repos/knative-sandbox/discovery/check-runs/7692116478",
  "html_url": "https://github.com/knative-sandbox/discovery/runs/7692116478?check_suite_focus=true",
  "details_url": "https://github.com/knative-sandbox/discovery/runs/7692116478?check_suite_focus=true",
  "status": "completed",
  "conclusion": "failure",
  "started_at": "2022-08-05T13:22:06Z",
  "completed_at": "2022-08-05T13:31:06Z",
  "output": {
    "title": "analyze / Analyze",
    "summary": "There are 1 failures, 0 warnings, and 0 notices.",
    "text": null,
    "annotations_count": 1,
    "annotations_url": "https://api.github.com/repos/knative-sandbox/discovery/check-runs/7692116478/annotations"
  },
  "check_suite": {
    "id": 7683866191
  },
  "app": {
    "id": 15368,
    "slug": "github-actions",
    "node_id": "MDM6QXBwMTUzNjg=",
    "owner": {
      "login": "github",
      "id": 9919,
      "node_id": "MDEyOk9yZ2FuaXphdGlvbjk5MTk=",
      "avatar_url": "https://avatars.githubusercontent.com/u/9919?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/github",
      "html_url": "https://github.com/github",
      "followers_url": "https://api.github.com/users/github/followers",
      "following_url": "https://api.github.com/users/github/following{/other_user}",
      "gists_url": "https://api.github.com/users/github/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/github/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/github/subscriptions",
      "organizations_url": "https://api.github.com/users/github/orgs",
      "repos_url": "https://api.github.com/users/github/repos",
      "events_url": "https://api.github.com/users/github/events{/privacy}",
      "received_events_url": "https://api.github.com/users/github/received_events",
      "type": "Organization",
      "site_admin": false
    },
    "name": "GitHub Actions",
    "description": "Automate your workflow from idea to production",
    "external_url": "https://help.github.com/en/actions",
    "html_url": "https://github.com/apps/github-actions",
    "created_at": "2018-07-30T09:30:17Z",
    "updated_at": "2019-12-10T19:04:12Z",
    "permissions": {
      "actions": "write",
      "administration": "read",
      "checks": "write",
      "contents": "write",
      "deployments": "write",
      "discussions": "write",
      "issues": "write",
      "merge_queues": "write",
      "metadata": "read",
      "packages": "write",
      "pages": "write",
      "pull_requests": "write",
      "repository_hooks": "write",
      "repository_projects": "write",
      "security_events": "write",
      "statuses": "write",
      "vulnerability_alerts": "read"
    },
    "events": [
      "branch_protection_rule",
      "check_run",
      "check_suite",
      "create",
      "delete",
      "deployment",
      "deployment_status",
      "discussion",
      "discussion_comment",
      "fork",
      "gollum",
      "issues",
      "issue_comment",
      "label",
      "merge_group",
      "milestone",
      "page_build",
      "project",
      "project_card",
      "project_column",
      "public",
      "pull_request",
      "pull_request_review",
      "pull_request_review_comment",
      "push",
      "registry_package",
      "release",
      "repository",
      "repository_dispatch",
      "status",
      "watch",
      "workflow_dispatch",
      "workflow_run"
    ]
  },
  "pull_requests": []
}
```

Added new checkrun via Check Runs API

```
 REDACTED  MCW0CDP3YY  ~  $  curl -XPOST  -H "Authorization:Bearer $TOKEN"   https://api.github.com/repos/knative-sandbox/discovery/check-runs   -d '{"name":"analyze / Analyze","head_sha":"8db7d8dd315cde9b446aece6175d914b010b841c","status":"completed","conclusion":"success","output":{"title":"Prow Override - analyze / Analyze","summary":"Prow has received an override for the analyze / Analyze checkrun"}'
{
  "id": 7730933836,
  "name": "analyze / Analyze",
  "node_id": "CR_kwDOEGSYGs8AAAABzMywTA",
  "head_sha": "8db7d8dd315cde9b446aece6175d914b010b841c",
  "external_id": "",
  "url": "https://api.github.com/repos/knative-sandbox/discovery/check-runs/7730933836",
  "html_url": "https://github.com/knative-sandbox/discovery/runs/7730933836",
  "details_url": "https://github.com/knative/infra",
  "status": "completed",
  "conclusion": "success",
  "started_at": "2022-08-08T17:23:33Z",
  "completed_at": "2022-08-08T17:23:33Z",
  "output": {
    "title": "Prow Override - analyze / Analyze",
    "summary": "Prow has received an override for the analyze / Analyze checkrun",
    "text": null,
    "annotations_count": 0,
    "annotations_url": "https://api.github.com/repos/knative-sandbox/discovery/check-runs/7730933836/annotations"
  },
  "check_suite": {
    "id": 7718870290
  },
  "app": {
    "id": 183726,
    "slug": "knative-prow",
    "node_id": "A_kwDOAnTdgM4AAs2u",
    "owner": {
      "login": "knative-prow-robot",
      "id": 41213312,
      "node_id": "MDQ6VXNlcjQxMjEzMzEy",
      "avatar_url": "https://avatars.githubusercontent.com/u/41213312?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/knative-prow-robot",
      "html_url": "https://github.com/knative-prow-robot",
      "followers_url": "https://api.github.com/users/knative-prow-robot/followers",
      "following_url": "https://api.github.com/users/knative-prow-robot/following{/other_user}",
      "gists_url": "https://api.github.com/users/knative-prow-robot/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/knative-prow-robot/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/knative-prow-robot/subscriptions",
      "organizations_url": "https://api.github.com/users/knative-prow-robot/orgs",
      "repos_url": "https://api.github.com/users/knative-prow-robot/repos",
      "events_url": "https://api.github.com/users/knative-prow-robot/events{/privacy}",
      "received_events_url": "https://api.github.com/users/knative-prow-robot/received_events",
      "type": "User",
      "site_admin": false
    },
    "name": "Knative Prow",
    "description": "This is the [Prow](https://github.com/kubernetes/test-infra/tree/master/prow) instance for running CI/CD jobs for [knative](https://github.com/knative) and [knative-sandbox](https://github.com/knative-sandbox) projects.",
    "external_url": "https://github.com/knative/infra",
    "html_url": "https://github.com/apps/knative-prow",
    "created_at": "2022-03-25T04:58:55Z",
    "updated_at": "2022-03-28T05:56:54Z",
    "permissions": {
      "actions": "write",
      "administration": "write",
      "checks": "write",
      "contents": "write",
      "dependabot_secrets": "read",
      "deployments": "read",
      "discussions": "read",
      "environments": "read",
      "issues": "write",
      "members": "write",
      "metadata": "read",
      "organization_administration": "write",
      "organization_dependabot_secrets": "read",
      "organization_events": "read",
      "organization_hooks": "read",
      "organization_plan": "read",
      "organization_projects": "admin",
      "organization_secrets": "read",
      "organization_self_hosted_runners": "read",
      "organization_user_blocking": "read",
      "packages": "read",
      "pages": "read",
      "pull_requests": "write",
      "repository_hooks": "read",
      "repository_projects": "admin",
      "secrets": "read",
      "secret_scanning_alerts": "read",
      "security_events": "read",
      "statuses": "write",
      "team_discussions": "read",
      "vulnerability_alerts": "read",
      "workflows": "write"
    },
    "events": [
      "branch_protection_rule",
      "check_run",
      "check_suite",
      "code_scanning_alert",
      "commit_comment",
      "create",
      "delete",
      "deployment",
      "deployment_review",
      "deployment_status",
      "deploy_key",
      "discussion",
      "discussion_comment",
      "fork",
      "gollum",
      "issues",
      "issue_comment",
      "label",
      "member",
      "membership",
      "merge_queue_entry",
      "milestone",
      "organization",
      "org_block",
      "project",
      "project_card",
      "project_column",
      "public",
      "pull_request",
      "pull_request_review",
      "pull_request_review_comment",
      "pull_request_review_thread",
      "push",
      "registry_package",
      "release",
      "repository",
      "repository_dispatch",
      "secret_scanning_alert",
      "star",
      "status",
      "team",
      "team_add",
      "watch",
      "workflow_dispatch",
      "workflow_job",
      "workflow_run"
    ]
  },
  "pull_requests": [

  ]
}
```

Reply:
PR has been merged

Both checkruns:

```
 REDACTED  MCW0CDP3YY  ~  $  gh api repos/knative-sandbox/discovery/commits/8db7d8dd315cde9b446aece6175d914b010b841c/check-runs | jq '.check_runs | .[] |select(.name=="analyze / Analyze") '
{
  "id": 7730933836,
  "name": "analyze / Analyze",
  "node_id": "CR_kwDOEGSYGs8AAAABzMywTA",
  "head_sha": "8db7d8dd315cde9b446aece6175d914b010b841c",
  "external_id": "",
  "url": "https://api.github.com/repos/knative-sandbox/discovery/check-runs/7730933836",
  "html_url": "https://github.com/knative-sandbox/discovery/runs/7730933836",
  "details_url": "https://github.com/knative/infra",
  "status": "completed",
  "conclusion": "success",
  "started_at": "2022-08-08T17:23:33Z",
  "completed_at": "2022-08-08T17:23:33Z",
  "output": {
    "title": "Prow Override - analyze / Analyze",
    "summary": "Prow has received an override for the analyze / Analyze checkrun",
    "text": null,
    "annotations_count": 0,
    "annotations_url": "https://api.github.com/repos/knative-sandbox/discovery/check-runs/7730933836/annotations"
  },
  "check_suite": {
    "id": 7718870290
  },
  "app": {
    "id": 183726,
    "slug": "knative-prow",
    "node_id": "A_kwDOAnTdgM4AAs2u",
    "owner": {
      "login": "knative-prow-robot",
      "id": 41213312,
      "node_id": "MDQ6VXNlcjQxMjEzMzEy",
      "avatar_url": "https://avatars.githubusercontent.com/u/41213312?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/knative-prow-robot",
      "html_url": "https://github.com/knative-prow-robot",
      "followers_url": "https://api.github.com/users/knative-prow-robot/followers",
      "following_url": "https://api.github.com/users/knative-prow-robot/following{/other_user}",
      "gists_url": "https://api.github.com/users/knative-prow-robot/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/knative-prow-robot/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/knative-prow-robot/subscriptions",
      "organizations_url": "https://api.github.com/users/knative-prow-robot/orgs",
      "repos_url": "https://api.github.com/users/knative-prow-robot/repos",
      "events_url": "https://api.github.com/users/knative-prow-robot/events{/privacy}",
      "received_events_url": "https://api.github.com/users/knative-prow-robot/received_events",
      "type": "User",
      "site_admin": false
    },
    "name": "Knative Prow",
    "description": "This is the [Prow](https://github.com/kubernetes/test-infra/tree/master/prow) instance for running CI/CD jobs for [knative](https://github.com/knative) and [knative-sandbox](https://github.com/knative-sandbox) projects.",
    "external_url": "https://github.com/knative/infra",
    "html_url": "https://github.com/apps/knative-prow",
    "created_at": "2022-03-25T04:58:55Z",
    "updated_at": "2022-03-28T05:56:54Z",
    "permissions": {
      "actions": "write",
      "administration": "write",
      "checks": "write",
      "contents": "write",
      "dependabot_secrets": "read",
      "deployments": "read",
      "discussions": "read",
      "environments": "read",
      "issues": "write",
      "members": "write",
      "metadata": "read",
      "organization_administration": "write",
      "organization_dependabot_secrets": "read",
      "organization_events": "read",
      "organization_hooks": "read",
      "organization_plan": "read",
      "organization_projects": "admin",
      "organization_secrets": "read",
      "organization_self_hosted_runners": "read",
      "organization_user_blocking": "read",
      "packages": "read",
      "pages": "read",
      "pull_requests": "write",
      "repository_hooks": "read",
      "repository_projects": "admin",
      "secrets": "read",
      "secret_scanning_alerts": "read",
      "security_events": "read",
      "statuses": "write",
      "team_discussions": "read",
      "vulnerability_alerts": "read",
      "workflows": "write"
    },
    "events": [
      "branch_protection_rule",
      "check_run",
      "check_suite",
      "code_scanning_alert",
      "commit_comment",
      "create",
      "delete",
      "deployment",
      "deployment_review",
      "deployment_status",
      "deploy_key",
      "discussion",
      "discussion_comment",
      "fork",
      "gollum",
      "issues",
      "issue_comment",
      "label",
      "member",
      "membership",
      "merge_queue_entry",
      "milestone",
      "organization",
      "org_block",
      "project",
      "project_card",
      "project_column",
      "public",
      "pull_request",
      "pull_request_review",
      "pull_request_review_comment",
      "pull_request_review_thread",
      "push",
      "registry_package",
      "release",
      "repository",
      "repository_dispatch",
      "secret_scanning_alert",
      "star",
      "status",
      "team",
      "team_add",
      "watch",
      "workflow_dispatch",
      "workflow_job",
      "workflow_run"
    ]
  },
  "pull_requests": []
}
{
  "id": 7692116478,
  "name": "analyze / Analyze",
  "node_id": "CR_kwDOEGSYGs8AAAABynxh_g",
  "head_sha": "8db7d8dd315cde9b446aece6175d914b010b841c",
  "external_id": "91148dad-8242-58a1-d104-2874ea6ecbb1",
  "url": "https://api.github.com/repos/knative-sandbox/discovery/check-runs/7692116478",
  "html_url": "https://github.com/knative-sandbox/discovery/runs/7692116478?check_suite_focus=true",
  "details_url": "https://github.com/knative-sandbox/discovery/runs/7692116478?check_suite_focus=true",
  "status": "completed",
  "conclusion": "failure",
  "started_at": "2022-08-05T13:22:06Z",
  "completed_at": "2022-08-05T13:31:06Z",
  "output": {
    "title": "analyze / Analyze",
    "summary": "There are 1 failures, 0 warnings, and 0 notices.",
    "text": null,
    "annotations_count": 1,
    "annotations_url": "https://api.github.com/repos/knative-sandbox/discovery/check-runs/7692116478/annotations"
  },
  "check_suite": {
    "id": 7683866191
  },
  "app": {
    "id": 15368,
    "slug": "github-actions",
    "node_id": "MDM6QXBwMTUzNjg=",
    "owner": {
      "login": "github",
      "id": 9919,
      "node_id": "MDEyOk9yZ2FuaXphdGlvbjk5MTk=",
      "avatar_url": "https://avatars.githubusercontent.com/u/9919?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/github",
      "html_url": "https://github.com/github",
      "followers_url": "https://api.github.com/users/github/followers",
      "following_url": "https://api.github.com/users/github/following{/other_user}",
      "gists_url": "https://api.github.com/users/github/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/github/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/github/subscriptions",
      "organizations_url": "https://api.github.com/users/github/orgs",
      "repos_url": "https://api.github.com/users/github/repos",
      "events_url": "https://api.github.com/users/github/events{/privacy}",
      "received_events_url": "https://api.github.com/users/github/received_events",
      "type": "Organization",
      "site_admin": false
    },
    "name": "GitHub Actions",
    "description": "Automate your workflow from idea to production",
    "external_url": "https://help.github.com/en/actions",
    "html_url": "https://github.com/apps/github-actions",
    "created_at": "2018-07-30T09:30:17Z",
    "updated_at": "2019-12-10T19:04:12Z",
    "permissions": {
      "actions": "write",
      "administration": "read",
      "checks": "write",
      "contents": "write",
      "deployments": "write",
      "discussions": "write",
      "issues": "write",
      "merge_queues": "write",
      "metadata": "read",
      "packages": "write",
      "pages": "write",
      "pull_requests": "write",
      "repository_hooks": "write",
      "repository_projects": "write",
      "security_events": "write",
      "statuses": "write",
      "vulnerability_alerts": "read"
    },
    "events": [
      "branch_protection_rule",
      "check_run",
      "check_suite",
      "create",
      "delete",
      "deployment",
      "deployment_status",
      "discussion",
      "discussion_comment",
      "fork",
      "gollum",
      "issues",
      "issue_comment",
      "label",
      "merge_group",
      "milestone",
      "page_build",
      "project",
      "project_card",
      "project_column",
      "public",
      "pull_request",
      "pull_request_review",
      "pull_request_review_comment",
      "push",
      "registry_package",
      "release",
      "repository",
      "repository_dispatch",
      "status",
      "watch",
      "workflow_dispatch",
      "workflow_run"
    ]
  },
  "pull_requests": []
}

```